core/src/main/java/org/apache/calcite/util/LazyReference.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.util;
+
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+/**
+ * This class can be used to lazily initialize an object.
+ *
+ * @param <T> Element Type
+ */
+public class LazyReference<T> {
+
+  private final AtomicReference<T> value = new AtomicReference<>();
+
+  public T getOrCompute(Supplier<T> supplier) {
```

Comment:
maybe the code is obvious, but I would still document the public API.
Since this is about concurrency, you should add in the comments a description of what this promises with respect to concurrent accesses.

Reply:
I added a description.